src/test/kotlin/org/jetbrains/intellij/IntelliJPluginSpec.kt

```diff
@@ -252,23 +161,147 @@ class IntelliJPluginSpec : IntelliJPluginSpecBase() {
 
-            assertTrue(compileClasspath.contains("copyright.jar"))
-            assertFalse(runtimeClasspath.contains("copyright.jar"))
-        }
+        assertContainsOnlySourceArtifacts(result,
+            "lib/src/go-openapi-src-goland-GO-212.5457.54-withSources-unzipped.com.jetbrains.plugins.jar " +
+                    "(unzipped.com.jetbrains.plugins:go:goland-GO-212.5457.54-withSources)",
+            "ideaIC-goland-GO-212.5457.54-withSources-sources.jar " +
+                    "(unzipped.com.jetbrains.plugins:go:goland-GO-212.5457.54-withSources)"
+        )
+    }
+
+    @Test
+    fun `add bundled zip plugin source artifacts from src directory when downloadSources = false`() {
+        buildFile.groovy("""
+            intellij {
+              type = 'GO'
+              version = '2021.2.4'
+              plugins = ['org.jetbrains.plugins.go']
+              downloadSources = false
+            }
+        """)
+
+        val result = printSourceArtifacts("unzipped.com.jetbrains.plugins:go:goland-GO")
+
+        assertContainsOnlySourceArtifacts(result,
+            "lib/src/go-openapi-src-goland-GO-212.5457.54-unzipped.com.jetbrains.plugins.jar " +
+                    "(unzipped.com.jetbrains.plugins:go:goland-GO-212.5457.54)"
+        )
+    }
+
+    @Test
+    fun `add external zip plugin source artifacts from src directory when downloadSources = true`() {
+        buildFile.groovy("""
+            intellij {
+              type = 'IC'
+              version = '2021.2.4'
+              plugins = ['org.jetbrains.plugins.go:212.5712.14'] // Go plugin is external for IC
+              downloadSources = true
+            }
+        """)
+
+        val result = printSourceArtifacts("unzipped.com.jetbrains.plugins:org.jetbrains.plugins.go")
+
+        assertContainsOnlySourceArtifacts(result,
+            "go/lib/src/go-openapi-src-212.5712.14-unzipped.com.jetbrains.plugins.jar " +
+                    "(unzipped.com.jetbrains.plugins:org.jetbrains.plugins.go:212.5712.14)"
+        )
+    }
+
+    @Test
+    fun `add external zip plugin source artifacts from src directory when downloadSources = false`() {
+        buildFile.groovy("""
+            intellij {
+              type = 'IC'
+              version = '2021.2.4'
+              plugins = ['org.jetbrains.plugins.go:212.5712.14'] // Go plugin is external for IC
+              downloadSources = false
+            }
+        """)
+        val result = printSourceArtifacts("unzipped.com.jetbrains.plugins:org.jetbrains.plugins.go")
+
+        assertContainsOnlySourceArtifacts(result,
+            "go/lib/src/go-openapi-src-212.5712.14-unzipped.com.jetbrains.plugins.jar " +
+                    "(unzipped.com.jetbrains.plugins:org.jetbrains.plugins.go:212.5712.14)"
+        )
+    }
+
+    @Test
+    fun `add bundled zip plugin source artifacts from src directory when localPath used`() {
+        val localPath = createLocalIdeIfNotExists(
```

Comment:
This downloads and unpacks IDEs next to the test Gradle home directory.
I created custom logic for downloading IDEs to not depend on the plugin logic (if reusing is even possible). It can be also used in other tests requiring local installations.